src/workspacesProvider.ts

```diff
@@ -14,34 +14,47 @@ export enum WorkspaceQuery {
 export class WorkspaceProvider implements vscode.TreeDataProvider<vscode.TreeItem> {
   private workspaces: WorkspaceTreeItem[] = []
   private agentWatchers: Record<WorkspaceAgent["id"], { dispose: () => void; metadata?: AgentMetadataEvent[] }> = {}
+  private timeout: NodeJS.Timeout | undefined
+  private visible = false
   private fetching = false
 
   constructor(
     private readonly getWorkspacesQuery: WorkspaceQuery,
     private readonly storage: Storage,
+    private readonly timerSeconds?: number,
   ) {
     this.fetchAndRefresh()
   }
 
-  // fetchAndRefresh fetches new workspaces then re-renders the entire tree.
-  // Trying to call this while already refreshing is a no-op and will return
+  // fetchAndRefresh fetches new workspaces, re-renders the entire tree, then
+  // keeps refreshing (if a timer length was provided) as long as the user is
+  // still logged in and no errors were encountered fetching workspaces.
+  // Calling this while already refreshing is a no-op and will return
   // immediately.
   async fetchAndRefresh() {
     if (this.fetching) {
       return
     }
     this.fetching = true
 
+    // TODO: It would be better to reuse these.
     Object.values(this.agentWatchers).forEach((watcher) => watcher.dispose())
 
+    // It is possible we called fetchAndRefresh() manually (through the button
+    // for example), in which case we might still have a pending refresh that
+    // needs to be cleared.
+    this.cancelPendingRefresh()
+
     try {
       this.workspaces = await this.fetch()
+      this.fetching = false
+      this.maybeScheduleRefresh()
     } catch (error) {
       this.workspaces = []
+      this.fetching = false
```

Comment:
Should we move this to a `finally` block?

Reply:
I really wanted to move it, but it has to be set before calling `this.maybeScheduleRefresh()` (since it checks whether a fetch is currently underway) so a `finally` would be too late.  So either way it would have to be set twice.

But it feels brittle.  I think maybe I refactor a bit to only set it once.